src/ray/gcs/redis_context-inl.h

```diff
@@ -0,0 +1,145 @@
+}
+
+template <typename RedisContextType, typename RedisConnectFunctionType>
+std::pair<Status, std::unique_ptr<RedisContextType, RedisContextDeleter>>
+ConnectWithRetries(const std::string &address,
+                   int port,
+                   const RedisConnectFunctionType &connect_function) {
+  RAY_LOG(INFO) << "Attempting to connect to address " << address << ":" << port << ".";
+  int connection_attempts = 0;
+  auto resp = ConnectWithoutRetries<RedisContextType>(address, port, connect_function);
+  auto status = resp.first;
+  while (!status.ok()) {
+    if (connection_attempts >= RayConfig::instance().redis_db_connect_retries()) {
+      RAY_LOG(FATAL) << RayConfig::instance().redis_db_connect_retries() << " attempts "
+                     << "to connect have all failed. Please check whether the"
+                     << " redis storage is alive or not. The last error message was: "
+                     << status.ToString();
+      break;
+    }
+    RAY_LOG_EVERY_MS(ERROR, 1000)
+        << "Failed to connect to Redis due to: " << status.ToString()
+        << ". Will retry in "
+        << RayConfig::instance().redis_db_connect_wait_milliseconds() << "ms.";
+
+    // Sleep for a little.
+    std::this_thread::sleep_for(std::chrono::milliseconds(
+        RayConfig::instance().redis_db_connect_wait_milliseconds()));
+    resp = ConnectWithoutRetries<RedisContextType>(address, port, connect_function);
+    status = resp.first;
+    connection_attempts += 1;
+  }
+  return resp;
+}
+
+template <typename ConnectType = redisAsyncContext>
+void RedisRequestContext::RedisResponseFn(struct redisAsyncContext *async_context,
+                                          void *raw_reply,
+                                          void *privdata) {
+  auto *request_cxt = static_cast<RedisRequestContext *>(privdata);
+  auto redis_reply = reinterpret_cast<redisReply *>(raw_reply);
+  // Error happened.
+  if (redis_reply == nullptr || redis_reply->type == REDIS_REPLY_ERROR) {
+    auto error_msg = redis_reply ? redis_reply->str : async_context->errstr;
+    RAY_LOG(ERROR) << "Redis request [" << absl::StrJoin(request_cxt->redis_cmds_, " ")
+                   << "]"
+                   << " failed due to error " << error_msg << ". "
+                   << request_cxt->pending_retries_ << " retries left.";
+    if (RayConfig::instance().enable_moved_redirect()) {
+      if (auto maybe_ip_port =
+              ParseIffMovedError(std::string(redis_reply->str, redis_reply->len));
+          maybe_ip_port.has_value()) {
+        const auto [ip, port] = maybe_ip_port.value();
+        auto resp = ConnectWithRetries<ConnectType>(ip.c_str(), port, redisAsyncConnect);
+        if (auto st = resp.first; !st.ok()) {
+          // We will ultimately return a MOVED error if we fail to reconnect.
+          RAY_LOG(ERROR) << "Failed to connect to the new leader " << ip << ":" << port;
+        } else {
+          request_cxt->redis_context_.reset(
```

Comment:
nit: make shared?

Reply:
btw, we don't want to always create the new context. if the leader changed, it probably will change for other keys since we have a 1 shard model.

Reply:
another question is that what's the consistence model for this? especially with the aws provided redis cluster?

Previously, the redirect is not taken care of because

- request ordering matters
- data consistence issues

If it's all read, data consistence is not a big issue, but if it's a write, previously, write will fail and crash and restarted. now write will succeed but we might still read the follower replica which end up with a stale data.

Reply:
Good point... I meant to change the one in RedisContext as well but you are right that this only affects the one in RedisRequestContext.

Reply:
For the consistency model: 
In order to get stale reads of a given key from the follower node, the read has to succeed to the follower node *after* a previous write to that node failed with MOVED and got redirected to the leader. 

For MOVED in particular, I thought that for any read or write that gets redirected, every subsequent read or write will also get redirected, so the read to the follower node should be redirected as well?

Reply:
also, apparently `make_unique` and `make_shared` don't allow custom deleters

Reply:
MemoryDB docs say that the primary node is strongly consistent:
https://docs.aws.amazon.com/memorydb/latest/devguide/consistency.html